src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -411,13 +411,51 @@ private static object CreateInstance([DynamicallyAccessedMembers(DynamicallyAcce
                 throw new InvalidOperationException(SR.Format(SR.Error_CannotActivateAbstractOrInterface, type));
             }
 
-            if (!type.IsValueType)
+            ConstructorInfo[] constructors = type.GetConstructors(DeclaredOnlyLookup);
+
+            bool hasParameterlessPublicConstructor =
+                constructors.Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
+
+            if (constructors.Length > 0 && !hasParameterlessPublicConstructor)
             {
-                bool hasDefaultConstructor = type.GetConstructors(DeclaredOnlyLookup).Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
-                if (!hasDefaultConstructor)
+                // if the only constructor is private, then throw
+                if (constructors.Length == 1 && !constructors[0].IsPublic)
                 {
                     throw new InvalidOperationException(SR.Format(SR.Error_MissingParameterlessConstructor, type));
                 }
+
+                // find the biggest constructor so that we can bind to the most parameters
+                ParameterInfo[]? chosenParameters = null;
+                ConstructorInfo? chosenConstructor = null;
+
+                for (int index = 0; index < constructors.Length; index++)
+                {
+                    ConstructorInfo constructor = constructors[index];
+                    ParameterInfo[] constructorParameters = constructor.GetParameters();
+
+                    if (!CanBindToTheseConstructorParameters(constructorParameters))
+                    {
+                        continue;
+                    }
+
+                    if (chosenParameters is null || constructorParameters.Length > chosenParameters.Length)
```

Comment:
Is it a problem that ties are broken based on constructor order which is inherently undefined? I wonder if it would be better if ties resulted in an exception.

In general, it feels like one of the most important things for this issue is to clearly define the constructor selection policy. Once selected, changing it would break compat. Just to summarize some of the considerations:

* Public default constructors are preferred for compat
* Are non-public constructors allowed (e. g. internal)?
* Are `in` and `ref` parameters allowed?
* Do we populate parameters with no default value and no configuration key (I think we shouldn't)?
* Do the available configuration keys matter for constructor selection (e. g. if I have just `"a": "2"` in my config and we're choosing between constructors `Options(int a)` and `Options(string x, string y)`, which one should it choose?
* What do we do if 2 constructors are equally eligible for selection?

One policy I might suggest is `select the public constructor with the most bindable parameters, where bindable means not ref or out, and either has a matching configuration key or has a default value. Prefer default constructors over any others for compat`.

A more conservative approach could be `Select the default constructor if there is one. Otherwise, if there is a single public constructor with no ref/out parameters use that. Otherwise, throw`. The advantage of this more conservative approach is that we can amend it later without breaking compat.

Reply:
We should look at what JSON does here. There’s prior art

Reply:
I took a look at the STJ source and [this article](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-immutability?pivots=dotnet-6-0). It chooses the default public parameterless constructor, or, if there is one, a constructor (doesn't have to be public) decorated with a `JsonConstructor` attribute. There's other ways to coerce the use of other constructors, such as using contract resolvers and the like.
STJ disallows non-public constructors.

Re. these points:

* Public default constructors are preferred for compat
Yes, I think that makes sense. It currently does this.

* Are non-public constructors allowed (e. g. internal)?
Currently, no. As above, this is consistent with System.Text.Json

* Are in and ref parameters allowed?
In STJ, both `ref` and `in` parameters are disallowed, resulting in
`Unhandled exception. System.ArgumentException: The type 'System.String&' may not be used as a type argument.
   at System.RuntimeType.ThrowIfTypeNeverValidGenericArgument(RuntimeType type)`
**Should we do the same? Or should we silently not consider the constructor?**

* Do we populate parameters with no default value and no configuration key (I think we shouldn't)?
Currently we do. I've added a test for this current behaviour (`CanBindImmutableClass_PicksLargestConstructorEvenIfItHasExtraneousParameters`)

* Do the available configuration keys matter for constructor selection (e. g. if I have just "a": "2" in my config and we're choosing between constructors Options(int a) and Options(string x, string y), which one should it choose?
The available keys are not considered. It currently picks the largest constructor in terms of parameters. As above, I added a test for this current behaviour (`CanBindImmutableClass_PicksLargestConstructorEvenIfItHasExtraneousParameters`)

* What do we do if 2 constructors are equally eligible for selection?
We pick first one. I added a test to cover this behaviour (`CanBindImmutableClass_PicksFirstOfAnyAmbiguousConstructors`)

You make some very valid points, and I think we should add constraints to avoid ambiguity and confusion. I think a model closer to STJ makes sense (except the use of an attribute):
* Use the default public constructor if it has one
* Use the only other parameterised constructor if there is one
* Disallow `in` and `ref`

Re. the parameterised constructor: I think we should leave this fairly tolerant with regards to the relationship between the parameters and the properties. It will allow config and classes to evolve separately, and will allow users to add 'synthetic' fields at bind time.  What do we think of this?

Reply:
Any thoughts on these suggestions? Making the behaviour closed to STJ seems like the right choice.

Reply:
Generally those look  like the right choices to me. Lets do it. Are we going to have an attribute to disambiguate ctors?

Reply:
I exercised some initiative and decided to go with the System.Text.Json way of binding constructors. This means that the behaviour now is this:

* Use the default public constructor if it has one
* Use the only other parameterised constructor if there is one
* Disallow in, out, and ref parameters on that constructor by throwing an `InvalidOperationException`
* Throw is there are multiple parameterised constructors

Reply:
How odd - I must've typed my last comment while you were typing yours! 

I think that if we don't consider non-public constructors, then we won't need an attribute to disambiguate any other constructor. That's because we use the parameterless constructor, if there is one, or the only other public parameterised constructor.
We _could_ consider an attributed constructor, but that decision is based more on the question of 'Should we allow non-public constructors?'

Reply:
> I think that if we don't consider non-public constructors, then we won't need an attribute to disambiguate any other constructor.

Structs always have a default constructor. What if you want to make config prefer your non-default constructor? Otherwise the following would always bind to `X = 0, Y = 0` because there's a default constructor with no parameters but the properties aren't settable.

```csharp
public struct Point
{
    public int X { get; }

    public int Y { get; }

    // An attribute would be nice here to tell Configuration not to use the default constructor
    public Point(int x, int y) => (X, Y) = (x, y);
}
```

Also, are we going to use default values for constructor parameters that are missing?

```csharp
public struct Point
{
    // ...
    public Point(int x, int y = 0) => (X, Y) = (x, y);
}
```

https://github.com/dotnet/runtime/blob/0e18cfda13a8136103c83442e1a649cdecc237b0/src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md#feature-behavior

Reply:
Good points. I'll add some more tests. If we go with an attribute, will that have to go through an API Proposal? And will that (should that) be part of another task and PR?

Reply:
Actually, binding to structs do work, and there is a test for it (`CanBindStructOptions`). The logic for constructors (currently) is:
* if it's **not** a value type && there are **no** constructors && it **doesn't** have a public parameterless constructor - then **throw an exception**
* if there are **multiple** constructors  && it **doesn't** have a public parameterless constructor, then **throw and exception**
* if it has just **one constructor** && **doesn't**  have a public parameterless constructor, then try to bind to constructor parameters
* otherwise, use `Activator.CreateInstance` with the default constructor (no parameter binding)